### PR TITLE
fix(utils): Handle more edgecases in `looseEqual()` util and add spec

### PR DESCRIPTION
**Pull request by jackmu95**
**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

This PR makes the `looseEqual()` util more bulletproof and adds an extra spec for it.

The previous implementation failed at following tests:

```bash
Expected: false
Received: true

  65 |     // Files with slightly different dates
> 66 |     expect(looseEqual(file1, file3)).toBe(false)
```

```bash
Expected: false
Received: true

  102 |     // Different regex
> 103 |     expect(looseEqual(rx1, rx3)).toBe(false)
```

```bash
Expected: false
Received: true

> 153 |     expect(looseEqual(obj1, date1)).toBe(false)
```

Performance should also be equal or slightly better because early exits are used wherever possible:
https://jsperf.com/looseequal-comparison

Shoutout to @tmorehouse for helping with the tests and handling all the edge cases.

### CODE REVIEW

File: test/unit/util/loose-equal.spec.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    // Dates with slightly different time (ms)
+    expect(looseEqual(date1, date3)).toBe(false)
+    // Dates with different year
+    expect(looseEqual(date1, date4)).toBe(false)
+  })
+
+  it('compares files correctly', () => {
+    const date1 = new Date(2019, 1, 2, 3, 4, 5, 6)
+    const date2 = new Date(2019, 1, 2, 3, 4, 5, 7)
+    const file1 = new File([''], 'filename.txt', { type: 'text/plain', lastModified: date1 })
```
Comment (posva) [inline test/unit/util/loose-equal.spec.js]: where is File coming from?

File: src/shared/util.js [hunk 2]
```diff
@@ -283,38 +288,42 @@ export function genStaticKeys (modules: Array<ModuleOptions>): string {
...
-        return a.getTime() === b.getTime()
-      } else if (!isArrayA && !isArrayB) {
-        const keysA = Object.keys(a)
-        const keysB = Object.keys(b)
-        return keysA.length === keysB.length && keysA.every(key => {
-          return looseEqual(a[key], b[key])
-        })
-      } else {
-        /* istanbul ignore next */
+  if (a === b) {
```
Comment (posva) [inline src/shared/util.js]: we like the one line 😉

File: packages/vue-server-renderer/build.dev.js [hunk 3]
```diff
@@ -41,10 +41,15 @@ function isPrimitive (value) {
   )
 }
 
+function isDate(obj) {
+  return obj instanceof Date
+}
```
Comment (posva) [inline packages/vue-server-renderer/build.dev.js]: make sure to remove any modifications to packages/* files

File: test/unit/util/loose-equal.spec.js (hunk 1 above)
Comment (jackmu95) [reply test/unit/util/loose-equal.spec.js]: What do you mean with "coming from"?

Comment (jackmu95) [reply test/unit/util/loose-equal.spec.js]: https://stackoverflow.com/a/26181292

Comment (posva) [reply test/unit/util/loose-equal.spec.js]: sorry, I meant why is this a test?

Comment (jackmu95) [reply test/unit/util/loose-equal.spec.js]: Because the previous implementation failed here.
It treated two completely different files as same.

This test covers that.

Comment (jackmu95) [reply test/unit/util/loose-equal.spec.js]: The main issue was to use `Object.keys()` to compare the keys across two objects.
It returns an empty array  for e.g. File objects.

File: test/unit/util/loose-equal.spec.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    const arr3 = [false, false, true]
+    const arr4 = [undefined, undefined, true]
+
+    expect(looseEqual(arr1, arr2)).toBe(true)
+    expect(looseEqual(arr2, arr1)).toBe(true)
+    expect(looseEqual(arr1, arr3)).toBe(false)
+    expect(looseEqual(arr3, arr1)).toBe(false)
+    expect(looseEqual(arr1, arr4)).toBe(true)
+    expect(looseEqual(arr4, arr1)).toBe(true)
+  })
```
Comment (tmorehouse) [inline test/unit/util/loose-equal.spec.js]: I would also add in the following test:

```js
    // This one is also sparse (missing index 1)
    const arr5 = []
    arr5[0] = arr5[2] = true

    // this one previously would return 'true', but shouldn't
    expect(looseEqual(arr1, arr5)).toBe(false)
    // this one previously would return 'false' (as it should)
    expect(looseEqual(arr5, arr1)).toBe(false)
```

**Comment by jackmu95** [comment]
@posva @yyx990803 Since the same changes landed in Vue v3 it would be awesome to have them in Vue v2.7 too.

Ref: https://github.com/vuejs/vue-next/pull/379
